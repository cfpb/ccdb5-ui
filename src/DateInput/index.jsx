/* eslint complexity: ["error", 7] */

import './DateInput.less'
import { bindAll, debounce, shortFormat } from '../utils'
import iconMap from '../iconMap';
import moment from 'moment'
import PropTypes from 'prop-types'
import React from 'react'

const FORMAT = 'MM-DD-YYYY'
const ONLY_VALID_SYMBOLS = /^[0-9/-]{1,10}$/
const HAS_4_DIGIT_YEAR = /^(\d{1,2})[-/](\d{1,2})[-/](\d{4})$/

// ----------------------------------------------------------------------------
// State Machine

const EMPTY = 'EMPTY'
const ACCUM = 'ACCUM'
const READY = 'READY'
const VALID = 'VALID'

const ERROR = 'ERROR'
const TOO_LOW = 'TOO_LOW'
const TOO_HIGH = 'TOO_HIGH'

// ----------------------------------------------------------------------------
// The Class

export default class DateInput extends React.Component {
  constructor( props ) {
    super( props )

    this.state = this._calculateState( props, props.value )

    bindAll( this, [ '_onClear', '_onChange', '_triggerCallbacks' ] )
    this._triggerCallbacks = props.debounceWait ?
      debounce( this._triggerCallbacks, props.debounceWait ) :
      this._triggerCallbacks
  }

  componentWillMount() {
    if ( this.state.message ) {
      this.props.onError( this.state.message, this.props.value )
    }
  }

  componentWillReceiveProps( nextProps ) {
    const state = this._calculateState( nextProps, nextProps.value )
    this.setState( state )
  }

  shouldComponentUpdate( nextProps, nextState ) {
    return JSON.stringify( this.state ) !== JSON.stringify( nextState )
  }

  componentDidUpdate() {
    this._triggerCallbacks()
  }

  render() {
    const placeholder = this.props.placeholder || 'mm/dd/yyyy'

    return (
      <div className="m-btn-inside-input">
        <input className={ this._inputClassName }
               onChange={ this._onChange }
               min={ this.props.min }
               max={ this.props.max }
               placeholder={ placeholder }
               size="15"
               value={ this.state.asText }
               {...this.props.textBoxProps}
        />
        { this.state.asText ?
          <button className="a-btn a-btn__link"
                  onClick={ this._onClear }>
              { iconMap.getIcon( 'delete' ) }
              <span className="u-visually-hidden">Clear</span>
          </button> :
          null
        }
      </div>
    )
  }

  // --------------------------------------------------------------------------
  // Properties

  get _inputClassName() {
    const style = [ 'a-text-input' ]
    if ( this.state.phase === ERROR ) {
      style.push( 'a-text-input__error' )
    }

    return style.join( ' ' )
  }

  // --------------------------------------------------------------------------
  // Event Handlers

  _onClear() {
    const newState = this._calculateState( this.props, '' )
    this.setState( newState )
  }

  _onChange( event ) {
    const v = event.target.value
    const newState = this._calculateState( this.props, v )
    this.setState( newState )
  }

  _triggerCallbacks() {
    switch ( this.state.phase ) {
      case EMPTY:
        if ( this.props.value ) {
          this.props.onDateEntered( null )
        }
        break

      case VALID:
        this.props.onDateEntered( this.state.asDate.toDate() )
        break

      case ERROR:
      case TOO_LOW:
      case TOO_HIGH:
        this.props.onError( this.state.message, this.state.asText )
        break

      default:
        this.props.onChange( this.state.asText )
        break
    }
  }

  // --------------------------------------------------------------------------
  // Validation

  _initialPhase( s ) {
    if ( !s ) {
      return EMPTY
    }

    let m = s.match( ONLY_VALID_SYMBOLS )
    if ( !m ) {
      return ERROR
    }

    if ( s.length < 8 ) {
      return ACCUM
    }

    m = s.match( HAS_4_DIGIT_YEAR )
    return m ? READY : ACCUM
  }

  _validateAsDate( props, d ) {
    if ( d.isValid() === false ) {
      return ERROR
    }

    if ( props.min && moment( d ).isBefore( props.min, 'day' ) ) {
      return TOO_LOW
    }

    if ( props.max && moment( d ).isAfter( props.max, 'day' ) ) {
      return TOO_HIGH
    }

    return VALID
  }

  _calculateState( props, v ) {
    const state = {
      asDate: null,
      asText: v,
      phase: this._initialPhase( v )
    }

    if ( state.phase === READY ) {
      state.asDate = moment( v, FORMAT )
      state.phase = this._validateAsDate( props, state.asDate )
    }

    state.message = this._buildErrorMessage( state.phase, props, v )

    return state
  }

  _buildErrorMessage( phase, props, v ) {
    switch ( phase ) {
      case ERROR:
        return `'${ v }' is not a valid date.`

      case TOO_LOW:
        return `'${ v }' must be greater than ` + shortFormat( props.min )

      case TOO_HIGH:
        return `'${ v }' must be less than ` + shortFormat( props.max )

      default:
        return ''
    }
  }
}

// --------------------------------------------------------------------------
// Meta

DateInput.propTypes = {
  debounceWait: PropTypes.number,
  max: PropTypes.instanceOf( Date ),
  min: PropTypes.instanceOf( Date ),
  onChange: PropTypes.func,
  onDateEntered: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  textBoxProps: PropTypes.object,
  value: PropTypes.string
}

/* eslint-disable no-empty-function */

DateInput.defaultProps = {
  debounceWait: 400,
  max: null,
  min: null,
  onChange: () => {},
  placeholder: '',
  textBoxProps: {},
  value: ''
}
