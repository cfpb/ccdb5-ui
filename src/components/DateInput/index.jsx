/* eslint complexity: ["error", 7] */

import './DateInput.less'
import { bindAll, debounce, shortFormat } from '../../utils'
import { DATE_VALIDATION_FORMAT } from '../../constants';
import dayjs from 'dayjs'
import dayjsCustomParseFormat from 'dayjs/plugin/customParseFormat'
import iconMap from '../iconMap'
import PropTypes from 'prop-types'
import React from 'react'

dayjs.extend( dayjsCustomParseFormat )

const ONLY_VALID_SYMBOLS = /^[0-9/]{1,10}$/
const HAS_4_DIGIT_YEAR = /^(\d{1,2})[-/](\d{1,2})[-/](\d{4})$/

// ----------------------------------------------------------------------------
// State Machine

const EMPTY = 'EMPTY'
const ACCUM = 'ACCUM'
const READY = 'READY'
const VALID = 'VALID'

const ERROR = 'ERROR'
const TOO_LOW = 'TOO_LOW'
const TOO_HIGH = 'TOO_HIGH'

// ----------------------------------------------------------------------------
// The Class

export default class DateInput extends React.Component {
  constructor( props ) {
    super( props )

    this.state = this._calculateState( props, props.value )

    bindAll( this, [ '_onClear', '_onChange', '_triggerCallbacks' ] )
    this._triggerCallbacks = props.debounceWait ?
      debounce( this._triggerCallbacks, props.debounceWait ) :
      this._triggerCallbacks
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillMount() {
    if ( this.state.message ) {
      this.props.onError( this.state.message, this.props.value )
    }
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillReceiveProps( nextProps ) {
    const state = this._calculateState( nextProps, nextProps.value )
    this.setState( state )
  }

  shouldComponentUpdate( nextProps, nextState ) {
    return JSON.stringify( this.state ) !== JSON.stringify( nextState )
  }

  componentDidUpdate() {
    this._triggerCallbacks()
  }

  render() {
    const placeholder = this.props.placeholder || DATE_VALIDATION_FORMAT

    return (
      <div className="m-btn-inside-input">
        <input className={ this._inputClassName }
               onChange={ this._onChange }
               min={ this.props.min }
               max={ this.props.max }
               placeholder={ placeholder }
               size="15"
               value={ this.state.asText }
               {...this.props.textBoxProps}
        />
        { this.state.asText ?
          <button className="a-btn a-btn__link"
                  onClick={ this._onClear }>
              { iconMap.getIcon( 'delete' ) }
              <span className="u-visually-hidden">Clear</span>
          </button> :
          null
        }
      </div>
    )
  }

  // --------------------------------------------------------------------------
  // Properties

  get _inputClassName() {
    const style = [ 'a-text-input' ]
    if ( this.state.phase === ERROR ) {
      style.push( 'a-text-input__error' )
    }

    return style.join( ' ' )
  }

  // --------------------------------------------------------------------------
  // Event Handlers

  _onClear() {
    const newState = this._calculateState( this.props, '' )
    this.setState( newState )
  }

  _onChange( event ) {
    const v = event.target.value.replace( '-', '/' )
    const newState = this._calculateState( this.props, v )
    this.setState( newState )
  }

  _triggerCallbacks() {
    switch ( this.state.phase ) {
      case EMPTY:
        if ( this.props.value ) {
          this.props.onDateEntered( null )
        }
        break

      case VALID:
        this.props.onDateEntered( this.state.asDate.toDate() )
        break

      case ERROR:
      case TOO_LOW:
      case TOO_HIGH:
        this.props.onError( this.state.message, this.state.asText )
        break

      default:
        this.props.onChange( this.state.asText )
        break
    }
  }

  // --------------------------------------------------------------------------
  // Validation

  _initialPhase( s ) {
    if ( !s ) {
      return EMPTY
    }

    let m = s.match( ONLY_VALID_SYMBOLS )
    if ( !m ) {
      return ERROR
    }

    if ( s.length < 8 ) {
      return ACCUM
    }

    m = s.match( HAS_4_DIGIT_YEAR )
    return m ? READY : ACCUM
  }

  _validateAsDate( props, d ) {
    if ( dayjs( d, DATE_VALIDATION_FORMAT, true ).isValid() === false ) {
      return ERROR
    }

    if ( props.min && dayjs( d ).isBefore( props.min, 'day' ) ) {
      return TOO_LOW
    }

    if ( props.max && dayjs( d ).isAfter( props.max, 'day' ) ) {
      return TOO_HIGH
    }

    return VALID
  }

  _calculateState( props, v ) {
    const state = {
      asDate: null,
      asText: v,
      phase: this._initialPhase( v )
    }

    if ( state.phase === READY ) {
      state.asDate = dayjs( v, DATE_VALIDATION_FORMAT, true )
      state.phase = this._validateAsDate( props, state.asDate )
    }

    state.message = this._buildErrorMessage( state.phase, props, v )

    return state
  }

  _buildErrorMessage( phase, props, v ) {
    switch ( phase ) {
      case ERROR:
        return `'${ v }' is not a valid date.`

      case TOO_LOW:
        return `'${ v }' must be greater than ` + shortFormat( props.min )

      case TOO_HIGH:
        return `'${ v }' must be less than ` + shortFormat( props.max )

      default:
        return ''
    }
  }
}

// --------------------------------------------------------------------------
// Meta

DateInput.propTypes = {
  debounceWait: PropTypes.number,
  max: PropTypes.instanceOf( Date ),
  min: PropTypes.instanceOf( Date ),
  onChange: PropTypes.func,
  onDateEntered: PropTypes.func.isRequired,
  onError: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  textBoxProps: PropTypes.object,
  value: PropTypes.string
}

/* eslint-disable no-empty-function */

DateInput.defaultProps = {
  debounceWait: 400,
  max: null,
  min: null,
  onChange: () => {},
  placeholder: '',
  textBoxProps: {},
  value: ''
}
