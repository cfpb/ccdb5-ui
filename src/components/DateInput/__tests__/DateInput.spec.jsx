import DateInput from '../index'
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme';

function setupEnzyme(initialProps={}) {
  const props = Object.assign({
    debounceWait: 0,
    onChange: jest.fn(),
    onDateEntered: jest.fn(),
    onError: jest.fn()
  }, initialProps)

  const target = mount(<DateInput {...props} />);

  return {
    props,
    target
  }
}

function setupSnapshot(value='') {
  return renderer.create(
    <DateInput debounceWait={0}
               onDateEntered={jest.fn()}
               onError={jest.fn()}
               value={value} />
  )
}


describe('component::DateInput', () => {
  describe('snapshots', () => {
    it('renders without crashing', () => {
      const target = setupSnapshot()
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders an error state', () => {
      const target = setupSnapshot('foo')
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('shows the clear button after text is entered', () => {
      const target = setupSnapshot('12')
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('date entry', () => {
    it('calls onDateEntered when the date is cleared', () => {
      const { target, props } = setupEnzyme({ value: '10/2/2015' })
      const input = target.find('input')
      input.simulate('change', { target: { value: '' }})
      expect(props.onDateEntered).toHaveBeenCalledWith(null)
    })

    it('does not call onDateEntered when value was already null', () => {
      const { target, props } = setupEnzyme()
      const input = target.find('input')
      input.simulate('change', { target: { value: '12/' }})
      expect(target.state('asText')).toEqual('12/')
      input.simulate('change', { target: { value: '' }})
      expect(target.state('asText')).toEqual('')
      expect(props.onDateEntered).not.toHaveBeenCalled()
    })

    it('calls onDateEntered when a valid date is entered', () => {
      const { target, props } = setupEnzyme()
      const input = target.find('input')
      input.simulate('change', { target: { value: '2/2/2010' }})
      const actual = props.onDateEntered.mock.calls[0][0]
      expect(actual).toEqual(expect.any(Date))
      expect(actual.getFullYear()).toEqual(2010)
      expect(actual.getMonth()).toEqual(2-1)
    })

    it('calls onError when the date is invalid', () => {
      const { target, props } = setupEnzyme()
      const input = target.find('input')
      input.simulate('change', { target: { value: '2/31/2012' }})
      expect(props.onError).toHaveBeenCalledWith(
        '\'2/31/2012\' is not a valid date.', '2/31/2012'
      )
    })

    it('calls onError when the date less than the minimum', () => {
      const { target, props } = setupEnzyme( { min: new Date( 2015, 0, 1 ) } )
      const input = target.find('input')
      input.simulate('change', { target: { value: '10/2/2010' } } )
      expect(props.onError).toHaveBeenCalledWith(
        '\'10/2/2010\' must be greater than 1/1/2015', '10/2/2010'
      )
    })

    it('calls onError when the date more than the maximum', () => {
      const { target, props } = setupEnzyme( { max: new Date( 2016, 0, 1 ) } )
      const input = target.find('input')
      input.simulate('change', { target: { value: '10/2/2017' } } )
      expect(props.onError).toHaveBeenCalledWith(
        '\'10/2/2017\' must be less than 1/1/2016', '10/2/2017'
      )
    })

    it('calls onChange for every other change', () => {
      const { target, props } = setupEnzyme()
      const input = target.find('input')
      input.simulate('change', { target: { value: '10/11/20' }})
      expect(props.onChange).toHaveBeenCalledWith( '10/11/20' )
    })
  })

  describe('clear button', () => {
    it('resets the control when clicked', () => {
      const { target, props } = setupEnzyme({ value: '10/2/2015' })
      const input = target.find('button')
      input.simulate('click')
      expect(target.state('asText')).toEqual('')
      expect(props.onDateEntered).toHaveBeenCalledWith(null)
    })
  })

  it('determines the correct starting mode from a string', () => {
    const parameterized = [
      ['1', 'ACCUM'],
      ['e', 'ERROR'],
      ['12', 'ACCUM'],
      ['12/', 'ACCUM'],
      ['12/12/201', 'ACCUM'],
      ['12/12/2012', 'READY'],
      ['2/12/e', 'ERROR']
    ]
    const { target } = setupEnzyme()

    parameterized.forEach(([fixture, expected]) => {
      const actual = target.instance()._initialPhase( fixture )
      expect(actual).toBe(expected)
    })
  })

  describe('no-op callbacks', () => {
    it('provides a no-op callback for onChange', () => {
      const actual = DateInput.defaultProps.onChange()
      expect(actual).toEqual(undefined)
    })
  })

  describe('componentWillReceiveProps', () => {
    it('validates the new properties', () => {
      const {target, props} = setupEnzyme()
      target.instance()._calculateState = jest.fn(() => ({}))
      target.setProps({value: '2016-01-01'})
      expect(target.instance()._calculateState).toHaveBeenCalledWith(
        expect.anything(), '2016-01-01'
      )
    })
  })
})

