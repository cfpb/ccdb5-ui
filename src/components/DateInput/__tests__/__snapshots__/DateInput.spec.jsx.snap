// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::DateInput snapshots renders an error state 1`] = `
<div
  className="m-btn-inside-input"
>
  <input
    className="a-text-input a-text-input__error"
    max={null}
    min={null}
    onChange={[Function]}
    placeholder="M/D/YYYY"
    size="15"
    value="foo"
  />
  <button
    className="a-btn a-btn__link"
    onClick={[Function]}
  >
    <svg
      className="cf-icon-svg"
      viewBox="0 0 718.9 1200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M451.4 613.7l248.1-248.1c25.6-25.1 26-66.3.8-91.9s-66.3-26-91.9-.8l-.8.8-248.1 248.1-248.1-248.1c-25.4-25.4-66.5-25.4-91.9 0s-25.4 66.5 0 91.9l248.1 248.1L19.5 861.8c-25.6 25.1-26 66.3-.8 91.9s66.3 26 91.9.8l.8-.8 248.1-248.1 248.1 248.1c25.4 25.4 66.5 25.4 91.9 0s25.4-66.5 0-91.9L451.4 613.7z"
      />
    </svg>
    <span
      className="u-visually-hidden"
    >
      Clear
    </span>
  </button>
</div>
`;

exports[`component::DateInput snapshots renders without crashing 1`] = `
<div
  className="m-btn-inside-input"
>
  <input
    className="a-text-input"
    max={null}
    min={null}
    onChange={[Function]}
    placeholder="M/D/YYYY"
    size="15"
    value=""
  />
</div>
`;

exports[`component::DateInput snapshots shows the clear button after text is entered 1`] = `
<div
  className="m-btn-inside-input"
>
  <input
    className="a-text-input"
    max={null}
    min={null}
    onChange={[Function]}
    placeholder="M/D/YYYY"
    size="15"
    value="12"
  />
  <button
    className="a-btn a-btn__link"
    onClick={[Function]}
  >
    <svg
      className="cf-icon-svg"
      viewBox="0 0 718.9 1200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M451.4 613.7l248.1-248.1c25.6-25.1 26-66.3.8-91.9s-66.3-26-91.9-.8l-.8.8-248.1 248.1-248.1-248.1c-25.4-25.4-66.5-25.4-91.9 0s-25.4 66.5 0 91.9l248.1 248.1L19.5 861.8c-25.6 25.1-26 66.3-.8 91.9s66.3 26 91.9.8l.8-.8 248.1-248.1 248.1 248.1c25.4 25.4 66.5 25.4 91.9 0s25.4-66.5 0-91.9L451.4 613.7z"
      />
    </svg>
    <span
      className="u-visually-hidden"
    >
      Clear
    </span>
  </button>
</div>
`;
